### PR TITLE
[cxxmodules] Don't overrite the DeserializationListener.

### PR DESCRIPTION
**Pull request by Teemperor**
With the module generation in rootcling Clang rlies on AST consumers
to do the module generation work for it. Right now this doesn't work
however with the interpreter, as we just overwrite the deserialization
listener that clang added which will cause strange errors during
the module generation (the most prompinent error is that the number
of recorded submodules will be incorrect, as this it the first thing
that Clang checks before writing a module and which is recorded by
an ASTDeserializationListener).

This patch just adds a multiplexer here that allows us to keep the
old listener while also adding the one we have.

### CODE REVIEW

File: interpreter/cling/lib/Interpreter/InterpreterCallbacks.cpp [hunk 1]
```diff
@@ -187,11 +201,23 @@ namespace cling {
...
+      // Create a new deserialization listener.
       m_DeserializationListener.
         reset(new InterpreterDeserializationListener(this));
-      Reader->setDeserializationListener(m_DeserializationListener.get());
+
+      // Wrap the deserialization listener in an MultiplexConsumer and then
+      // combine it with the existing Consumer.
+      // FIXME: Maybe it's better to make MultiplexASTDeserializationListener
+      // public instead. See also: https://reviews.llvm.org/D37475
+      std::unique_ptr<DeserializationListenerWrapper> wrapper(new DeserializationListenerWrapper(m_DeserializationListener.get()));
```
Comment (Axel-Naumann) [inline interpreter/cling/lib/Interpreter/InterpreterCallbacks.cpp]: Does that fit into cling's / clang's 80 char line width limit?

File: interpreter/cling/lib/Interpreter/InterpreterCallbacks.cpp [hunk 2]
```diff
@@ -78,6 +79,19 @@ namespace cling {
     }
   };
 
+  /// \brief Wraps an ASTDeserializationListener in an ASTConsumer so that
+  /// it can be used with a MultiplexConsumer.
+  class DeserializationListenerWrapper : public ASTConsumer {
```
Comment (Axel-Naumann) [inline interpreter/cling/lib/Interpreter/InterpreterCallbacks.cpp]: should this be in an unnamed namespace?

File: interpreter/cling/lib/Interpreter/InterpreterCallbacks.cpp [hunk 3]
```diff
@@ -78,6 +79,19 @@ namespace cling {
...
+    ASTDeserializationListener *m_Listener;
+  public:
+    DeserializationListenerWrapper(ASTDeserializationListener *Listener)
```
Comment (Axel-Naumann) [inline interpreter/cling/lib/Interpreter/InterpreterCallbacks.cpp]: Should this own the `ASTDeserializationListener`?

**Comment by vgvassilev** [comment]
Could you add a test testing this under `cling/test`?

File: interpreter/cling/lib/Interpreter/InterpreterCallbacks.cpp (hunk 1 above)
Comment (Teemperor) [reply interpreter/cling/lib/Interpreter/InterpreterCallbacks.cpp]: Nope, that's why Travis is red :) Fixed!

File: interpreter/cling/lib/Interpreter/InterpreterCallbacks.cpp (hunk 2 above)
Comment (Teemperor) [reply interpreter/cling/lib/Interpreter/InterpreterCallbacks.cpp]: Fixed!

**Pull request merged**